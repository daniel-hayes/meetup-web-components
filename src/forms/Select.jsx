// @flow
import * as React from 'react';
import cx from 'classnames';

import Icon from '../media/Icon';
import withErrorList from '../utils/components/withErrorList';

type Props = React.ElementConfig<HTMLSelectElement> & {
	/** Additional class name/s to add to the `<select/>` element  */
	className?: string,

	/** The `name` attribute for the input, and associates it with the `<label />` */
	name: string, // required - will be used as 'id' as well

	/** Required - this component is stateless. A callback that happens when the input changes */
	onChange: (e: SyntheticInputEvent<*>) => void,

	/** Supplied by withErrorList */
	error?: React.Node,

	/** An additional piece of helpful info rendered with the field */
	helperText?: string,

	/** The class name/s to add to the `<label />` element */
	labelClassName?: string,

	/** The class name/s to add to the div element wrapping select input + icon */
	selectClassName?: string,

	/** What we render into the input's `<label />` */
	label?: string,

	/** What to render in order to indicate the field is required. Supply a string for a custom error message */
	required?: boolean | string,
};

/*
 * MWC custom <select> component
 */
export const SelectInput = (props: Props) => {
	const {
		className,
		id,
		labelClassName,
		label,
		name,
		error,
		helperText,
		required,
		selectClassName,
		...other
	} = props;

	const classNames = {
		label: cx(
			'label--field',
			{ 'label--required': required, 'flush--bottom': helperText },
			labelClassName
		),
		field: cx(
			'select--reset span--100 padding--selectArrow',
			{ 'field--error': error },
			className
		),
		helperText: cx('helperTextContainer', { required }),
	};
	const requiredText = typeof required === 'string' ? required : '*';

	return (
		<div className="inputContainer">
			{label && (
				<label
					className={classNames.label}
					htmlFor={name}
					data-requiredtext={requiredText}
				>
					{label}
				</label>
			)}
			{helperText && <div className={classNames.helperText}>{helperText}</div>}
			<div className={`selectWrapper ${selectClassName || ''}`}>
				<select
					name={name}
					id={id || name}
					required={Boolean(required)}
					className={classNames.field}
					{...other}
				/>
				<Icon className="select-customArrow" shape="chevron-down" size="xs" />
			</div>
		</div>
	);
};

const SelectWithErrors = withErrorList(SelectInput);
SelectWithErrors.displayName = 'Select';

export default SelectWithErrors;
