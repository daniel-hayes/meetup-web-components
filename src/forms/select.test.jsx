import React from 'react';
import { shallow } from 'enzyme';
import { SelectInput } from './Select';

const testOptions = (
	<React.Fragment>
		<option value="1">One</option>,
		<option value="2">Two</option>,
		<option value="3">Three</option>,
		<option value="4" disabled>
			Four
		</option>,
	</React.Fragment>
);

const BasicSelect = (
	<SelectInput label="Test select" name="testSelect" required="yo check it">
		{testOptions}
	</SelectInput>
);
describe('SelectInput basic', () => {
	const component = shallow(BasicSelect);

	it('renders into the DOM', () => {
		expect(component).toMatchSnapshot();
	});
});
