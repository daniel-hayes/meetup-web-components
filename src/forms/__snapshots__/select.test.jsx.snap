// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectInput basic renders into the DOM 1`] = `
<Select
  id="testSelect"
  label="Test select"
  name="testSelect"
  requiredText="yo check it"
>
  <option
    value="1"
  >
    One
  </option>
  ,
  <option
    value="2"
  >
    Two
  </option>
  ,
  <option
    value="3"
  >
    Three
  </option>
  ,
  <option
    disabled={true}
    value="4"
  >
    Four
  </option>
  ,
</Select>
`;
